scipy/optimize/optimize.py

```diff
@@ -2606,12 +2607,35 @@ def _scalarfunc(*params):
         grid = grid[0]
         xmin = xmin[0]
     if callable(finish):
-        vals = finish(func, xmin, args=args, full_output=1, disp=disp)
-        xmin = vals[0]
-        Jmin = vals[1]
-        if vals[-1] > 0:
+        # set up kwargs for `finish` function
+        finish_args = getargspec(finish).args
+        kwargs = dict()
+        if 'full_output' in finish_args:
+            kwargs['full_output'] = 1
+        if 'disp' in finish_args:
+            kwargs['disp'] = disp
+        elif 'options' in finish_args:
+            # pass 'disp' as `options`
+            # (e.g. if `finish` is `minimize`)
+            kwargs['options'] = {'disp': disp}
+
+        # run minimizer
+        res = finish(func, xmin, args=args, **kwargs)
+
+        if isinstance(res, OptimizeResult):
+            xmin = res.x
+            Jmin = res.fun
+            success = res.success
+        else:
+            xmin = res[0]
+            Jmin = res[1]
+            success = res[-1] == 0
+        if not success:
             if disp:
-                print("Warning: Final optimization did not succeed")
+                print("Warning: Either final optimization did not succeed " +
+                      "or `finish` does not return `statuscode` as its last " +
+                      "argument.")
+
```

Comment:
`+` are superfluous here. In Python, you can just concatenate strings like this `('abc' 'def') == 'abcdef'`.